File: src/microsoft-agent-framewokr-agt-integration/dotnet/AgtMaf/Analyzers.cs
using System.Text.RegularExpressions;
using AgentGovernance;
using YamlDotNet.Serialization;
using YamlDotNet.Serialization.NamingConventions;

namespace AgtMaf;

/// <summary>
/// Loads an AGT policy file into a <see cref="GovernanceKernel"/> and also parses it
/// for each rule's custom <c>message</c>. The .NET PolicyDecision surfaces the matched
/// rule NAME but not its message, so we read the messages ourselves to show friendly
/// reasons (and to detect the "requires human approval" escalation marker).
/// </summary>
public sealed class PolicyBundle
{
    public GovernanceKernel Kernel { get; }
    public IReadOnlyDictionary<string, string> RuleMessages { get; }

    private PolicyBundle(GovernanceKernel kernel, IReadOnlyDictionary<string, string> messages)
    {
        Kernel = kernel;
        RuleMessages = messages;
    }

    public string? MessageFor(string? ruleName) =>
        ruleName is not null && RuleMessages.TryGetValue(ruleName, out var m) ? m : null;

    public static PolicyBundle FromFile(string path)
    {
        var yaml = File.ReadAllText(path);

        var kernel = new GovernanceKernel();
        kernel.LoadPolicyFromYaml(yaml);

        var messages = new Dictionary<string, string>(StringComparer.Ordinal);
        try
        {
            var deserializer = new DeserializerBuilder()
                .WithNamingConvention(CamelCaseNamingConvention.Instance)
                .IgnoreUnmatchedProperties()
                .Build();
            var doc = deserializer.Deserialize<PolicyFileModel>(yaml);
            foreach (var rule in doc?.Rules ?? [])
            {
                if (!string.IsNullOrWhiteSpace(rule.Name) && !string.IsNullOrWhiteSpace(rule.Message))
                {
                    messages[rule.Name!] = rule.Message!;
                }
            }
        }
        catch
        {
            // Friendly messages are best-effort; the kernel decision is the source of truth.
        }

        return new PolicyBundle(kernel, messages);
    }

    private sealed class PolicyFileModel
    {
        public List<RuleModel>? Rules { get; set; }
    }

    private sealed class RuleModel
    {
        public string? Name { get; set; }
        public string? Message { get; set; }
    }
}

/// <summary>Deterministic signals extracted from a prompt by static analysis.</summary>
public sealed record PromptFeatures(
    bool ContainsPii,
    bool ContainsSecret,
    bool ContainsInjection,
    IReadOnlyList<string> InjectionMarkers);

/// <summary>
/// Govern an outbound prompt with deterministic C# static analysis plus a real AGT
/// policy. The regex / keyword detection runs here (the .NET policy engine has no regex
/// operator); the resulting boolean signals are fed into the policy, which decides.
/// </summary>
public sealed class StaticPromptAnalyzer
{
    private static readonly Regex SsnRe = new(@"\b\d{3}-\d{2}-\d{4}\b", RegexOptions.Compiled);
    private static readonly Regex CreditCardRe = new(@"\b(?:\d[ -]?){15,16}\b", RegexOptions.Compiled);

    private static readonly Regex SecretRe = new(
        @"\b(AKIA[0-9A-Z]{16}|sk-[A-Za-z0-9]{20,}|ghp_[A-Za-z0-9]{20,}|xox[baprs]-[A-Za-z0-9-]{10,}|eyJ[A-Za-z0-9_-]{10,}\.[A-Za-z0-9_-]{10,})\b",
        RegexOptions.Compiled | RegexOptions.IgnoreCase);

    private static readonly Regex SecretKeywordRe = new(
        @"\b(password|passwd|api[_ -]?key|secret[_ -]?key|connection[_ -]?string|bearer\s+token)\b",
        RegexOptions.Compiled | RegexOptions.IgnoreCase);

    private static readonly Regex[] InjectionPatterns =
    [
        new(@"ignore (all |any |the )?(previous|prior|above) instructions", RegexOptions.Compiled | RegexOptions.IgnoreCase),
        new(@"disregard (the|all|your) (above|previous|prior|system)", RegexOptions.Compiled | RegexOptions.IgnoreCase),
        new(@"reveal (your|the) (system|developer) prompt", RegexOptions.Compiled | RegexOptions.IgnoreCase),
        new(@"you are now (in )?(developer|dan|jailbreak) mode", RegexOptions.Compiled | RegexOptions.IgnoreCase),
        new(@"\bjailbreak\b", RegexOptions.Compiled | RegexOptions.IgnoreCase),
        new(@"exfiltrate|leak (the )?(data|secrets|credentials)", RegexOptions.Compiled | RegexOptions.IgnoreCase),
        new(@"pretend (to be|you are)|act as (an?|the) (unrestricted|uncensored)", RegexOptions.Compiled | RegexOptions.IgnoreCase),
    ];

    private readonly PolicyBundle _policy;
    private readonly string _agentId;

    public StaticPromptAnalyzer(string policyPath, string agentId)
    {
        _policy = PolicyBundle.FromFile(policyPath);
        _agentId = agentId;
    }

    public static PromptFeatures ExtractFeatures(string prompt)
    {
        var markers = new List<string>();
        foreach (var pat in InjectionPatterns)
        {
            var m = pat.Match(prompt);
            if (m.Success)
            {
                markers.Add(m.Value);
            }
        }
        var pii = SsnRe.IsMatch(prompt) || CreditCardRe.IsMatch(prompt);
        var secret = SecretRe.IsMatch(prompt) || SecretKeywordRe.IsMatch(prompt);
        return new PromptFeatures(pii, secret, markers.Count > 0, markers);
    }

    public (GovernanceDecision Decision, PromptFeatures Features) Analyze(string prompt)
    {
        var features = ExtractFeatures(prompt);

        // The deterministic signals are the evaluation context the policy decides on.
        var context = new Dictionary<string, object>(StringComparer.OrdinalIgnoreCase)
        {
            ["contains_pii"] = features.ContainsPii,
            ["contains_secret"] = features.ContainsSecret,
            ["contains_injection"] = features.ContainsInjection,
            ["injection_markers"] = features.InjectionMarkers.Count,
        };

        var decision = _policy.Kernel.PolicyEngine.Evaluate(_agentId, context);

        var summary = prompt.Length > 58 ? prompt[..57] + "..." : prompt;
        var gov = GovernanceDecision.Create(
            decision.Allowed,
            decision.MatchedRule,
            _policy.MessageFor(decision.MatchedRule),
            decision.Reason,
            layer: "prompt",
            target: summary);
        return (gov, features);
    }
}

/// <summary>
/// Govern an outbound tool call with a real AGT capability + argument-boundary policy.
/// This is the heart of the scenario: the same agent-level policy decides allow / deny /
/// escalate purely from the tool name AND its argument values.
/// </summary>
public sealed class ToolCallAnalyzer
{
    private readonly PolicyBundle _policy;
    private readonly string _agentId;

    public ToolCallAnalyzer(string policyPath, string agentId)
    {
        _policy = PolicyBundle.FromFile(policyPath);
        _agentId = agentId;
    }

    public GovernanceDecision Analyze(string toolName, IReadOnlyDictionary<string, object?> arguments)
    {
        // The kernel merges these args into the evaluation context, so policy conditions
        // can reference argument values directly (e.g. `amount > 50000`).
        var args = new Dictionary<string, object>(StringComparer.OrdinalIgnoreCase);
        foreach (var (key, value) in arguments)
        {
            if (value is not null)
            {
                args[key] = value;
            }
        }

        var result = _policy.Kernel.EvaluateToolCall(_agentId, toolName, args);
        return GovernanceDecision.Create(
            result.Allowed,
            result.PolicyDecision?.MatchedRule,
            _policy.MessageFor(result.PolicyDecision?.MatchedRule),
            result.Reason,
            layer: "tool",
            target: toolName);
    }
}

File: src/agent-governance-toolkit/microsoft-agent-framework-demos/python/agt_maf/analyzers.py
"""Deterministic static analysis backed by the Agent Governance Toolkit (AGT).

Two analyzers live here:

* :class:`StaticPromptAnalyzer` — extracts deterministic features from an outbound
  prompt (PII, secrets, prompt-injection signatures) and evaluates them with a real
  AGT :class:`~agent_os.policies.PolicyEvaluator`. This is the analyzer the prompt
  governance middleware uses on every user turn.

* :class:`ToolCallAnalyzer` — evaluates an outbound tool call (the tool name plus its
  flattened arguments) with a real AGT ``PolicyEvaluator``. This is the analyzer the
  tool governance middleware uses on every tool call.

* :class:`PromptHardeningAuditor` — a *build-time* static analysis that grades an
  agent's **system instructions** for prompt-injection resistance across AGT's
  12-vector PromptDefense model (OWASP LLM Top-10 aligned). Use this to check that
  your agent's own instructions are hardened — it is not a per-request attack filter.

Everything here is deterministic: the same input always yields the same decision,
which is exactly what makes it a usable control surface for autonomous agents.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from agent_os.policies import PolicyEvaluator
from agent_os.policies.schema import PolicyDocument

from .model import GovernanceDecision

# ---------------------------------------------------------------------------
# Deterministic feature extraction for prompts
# ---------------------------------------------------------------------------

# Well-known sensitive-data signatures.
_SSN_RE = re.compile(r"\b\d{3}-\d{2}-\d{4}\b")
_CREDIT_CARD_RE = re.compile(r"\b(?:\d[ -]?){15,16}\b")
_SECRET_RE = re.compile(
    r"(?i)\b("
    r"AKIA[0-9A-Z]{16}"  # AWS access key id
    r"|sk-[A-Za-z0-9]{20,}"  # OpenAI-style secret key
    r"|ghp_[A-Za-z0-9]{20,}"  # GitHub PAT
    r"|xox[baprs]-[A-Za-z0-9-]{10,}"  # Slack token
    r"|eyJ[A-Za-z0-9_-]{10,}\.[A-Za-z0-9_-]{10,}"  # JWT
    r")\b"
)
_SECRET_KEYWORD_RE = re.compile(
    r"(?i)\b(password|passwd|api[_ -]?key|secret[_ -]?key|connection[_ -]?string|bearer\s+token)\b"
)

# Prompt-injection / jailbreak signatures (deterministic phrase list).
_INJECTION_PATTERNS: tuple[re.Pattern[str], ...] = (
    re.compile(r"(?i)ignore (all |any |the )?(previous|prior|above) instructions"),
    re.compile(r"(?i)disregard (the|all|your) (above|previous|prior|system)"),
    re.compile(r"(?i)reveal (your|the) (system|developer) prompt"),
    re.compile(r"(?i)you are now (in )?(developer|dan|jailbreak) mode"),
    re.compile(r"(?i)\bjailbreak\b"),
    re.compile(r"(?i)exfiltrate|leak (the )?(data|secrets|credentials)"),
    re.compile(r"(?i)pretend (to be|you are)|act as (an?|the) (unrestricted|uncensored)"),
)


@dataclass
class PromptFeatures:
    """Deterministic signals extracted from a prompt by static analysis."""

    prompt_length: int
    contains_pii: bool
    contains_secret: bool
    injection_marker_count: int
    injection_markers: list[str]

    def as_context(self, prompt: str) -> dict[str, Any]:
        """Flatten into the context dict an AGT policy evaluates against."""
        return {
            "input_text": prompt,
            "prompt_length": self.prompt_length,
            "contains_pii": self.contains_pii,
            "contains_secret": self.contains_secret,
            "injection_marker_count": self.injection_marker_count,
        }


def extract_prompt_features(prompt: str) -> PromptFeatures:
    """Run deterministic static analysis over a prompt and return its signals."""
    markers: list[str] = []
    for pat in _INJECTION_PATTERNS:
        m = pat.search(prompt)
        if m:
            markers.append(m.group(0))
    contains_pii = bool(_SSN_RE.search(prompt) or _CREDIT_CARD_RE.search(prompt))
    contains_secret = bool(_SECRET_RE.search(prompt) or _SECRET_KEYWORD_RE.search(prompt))
    return PromptFeatures(
        prompt_length=len(prompt),
        contains_pii=contains_pii,
        contains_secret=contains_secret,
        injection_marker_count=len(markers),
        injection_markers=markers,
    )


# ---------------------------------------------------------------------------
# Prompt analyzer (AGT PolicyEvaluator over the prompt + its derived features)
# ---------------------------------------------------------------------------


class StaticPromptAnalyzer:
    """Govern an outbound prompt with deterministic features + a real AGT policy."""

    def __init__(self, policy: str | Path | PolicyDocument | PolicyEvaluator) -> None:
        self.evaluator = _as_evaluator(policy)

    def analyze(self, prompt: str) -> tuple[GovernanceDecision, PromptFeatures]:
        features = extract_prompt_features(prompt)
        decision = self.evaluator.evaluate(features.as_context(prompt))
        summary = (prompt[:57] + "…") if len(prompt) > 58 else prompt
        gov = GovernanceDecision.from_policy_decision(
            decision,
            layer="prompt",
            target=summary,
            features={
                "contains_pii": features.contains_pii,
                "contains_secret": features.contains_secret,
                "injection_markers": features.injection_markers,
            },
        )
        return gov, features


# ---------------------------------------------------------------------------
# Tool-call analyzer (AGT PolicyEvaluator over tool name + flattened arguments)
# ---------------------------------------------------------------------------


class ToolCallAnalyzer:
    """Govern an outbound tool call with a real AGT policy."""

    def __init__(self, policy: str | Path | PolicyDocument | PolicyEvaluator) -> None:
        self.evaluator = _as_evaluator(policy)

    def analyze(self, tool_name: str, arguments: dict[str, Any]) -> GovernanceDecision:
        context: dict[str, Any] = {"tool_name": tool_name}
        # Flatten the arguments into the evaluation context so a single, agent-level
        # policy can reason about tool-call ARGUMENTS — not just tool names. Each
        # argument is exposed two ways:
        #
        #   * flat        e.g. ``amount``                 -> a rule applies to *any*
        #                                                    tool that takes ``amount``
        #   * namespaced  e.g. ``transfer_budget.amount`` -> a rule applies to *only*
        #                                                    that tool's argument
        #
        # The namespaced key is what makes precise, per-tool argument boundaries
        # possible from one central policy (the evaluator does a flat key lookup, so
        # ``transfer_budget.amount`` is just a literal context key it can match on).
        for key, value in (arguments or {}).items():
            context[key] = value
            context[f"{tool_name}.{key}"] = value
        decision = self.evaluator.evaluate(context)
        return GovernanceDecision.from_policy_decision(
            decision,
            layer="tool",
            target=tool_name,
            features={"arguments": dict(arguments or {})},
        )


# ---------------------------------------------------------------------------
# Build-time prompt hardening audit (AGT PromptDefense — 12 OWASP vectors)
# ---------------------------------------------------------------------------


@dataclass
class HardeningFinding:
    vector_id: str
    name: str
    owasp: str
    defended: bool
    severity: str
    evidence: str


@dataclass
class HardeningReport:
    grade: str
    score: int
    findings: list[HardeningFinding]

    @property
    def defended_count(self) -> int:
        return sum(1 for f in self.findings if f.defended)

    @property
    def gaps(self) -> list[HardeningFinding]:
        return [f for f in self.findings if not f.defended]


class PromptHardeningAuditor:
    """Statically grade an agent's *system instructions* for injection resistance.

    This wraps the Agent Governance Toolkit's ``PromptDefenseEvaluator`` (shipped in
    the ``agent_compliance`` package). It scans text for defensive guardrail patterns
    across 12 OWASP-LLM-aligned vectors and assigns a letter grade. Use it at build
    time to confirm an agent's own instructions are hardened — higher grade = more
    injection-resistant instructions.
    """

    def __init__(self, min_grade: str = "B") -> None:
        from agent_compliance import PromptDefenseConfig, PromptDefenseEvaluator

        self._evaluator = PromptDefenseEvaluator(PromptDefenseConfig(min_grade=min_grade))
        self.min_grade = min_grade

    def audit(self, instructions: str) -> HardeningReport:
        report = self._evaluator.evaluate(instructions)
        findings = [
            HardeningFinding(
                vector_id=f.vector_id,
                name=f.name,
                owasp=f.owasp,
                defended=f.defended,
                severity=f.severity,
                evidence=f.evidence,
            )
            for f in report.findings
        ]
        return HardeningReport(grade=report.grade, score=report.score, findings=findings)


# ---------------------------------------------------------------------------
# helpers
# ---------------------------------------------------------------------------


def _as_evaluator(policy: str | Path | PolicyDocument | PolicyEvaluator) -> PolicyEvaluator:
    if isinstance(policy, PolicyEvaluator):
        return policy
    if isinstance(policy, PolicyDocument):
        return PolicyEvaluator(policies=[policy])
    # Load the YAML ourselves as UTF-8 and validate via the pydantic model. This is
    # equivalent to ``PolicyDocument.from_yaml`` but encoding-correct on every OS:
    # some installed AGT builds open policy files with the platform default encoding
    # (cp1252 on Windows), which fails on any non-ASCII byte in the file.
    import yaml

    data = yaml.safe_load(Path(policy).read_text(encoding="utf-8"))
    doc = PolicyDocument.model_validate(data)
    return PolicyEvaluator(policies=[doc])
